pkg/namespacestore/validation.go

```diff
@@ -0,0 +1,160 @@
+	switch nsStore.Spec.Type {
+
+	case nbv1.NSStoreTypeNSFS:
+		return validateNsStoreNSFS(nsStore)
+
+	case nbv1.NSStoreTypeAWSS3:
+		return nil
+
+	case nbv1.NSStoreTypeS3Compatible:
+		return validateNsStoreS3Compatible(nsStore)
+
+	case nbv1.NSStoreTypeIBMCos:
+		return validateNsStoreIBMCos(nsStore)
+
+	case nbv1.NSStoreTypeAzureBlob:
+		return nil
+
+	default:
+		return util.NewPersistentError("ConfigurationError",
+			fmt.Sprintf("Invalid namespace store type %q", nsStore.Spec.Type))
+	}
+}
+
+// validateNamespaceStoreNSFS validates namespacestore nsfs type configuration
+func validateNsStoreNSFS(nsStore *nbv1.NamespaceStore) error {
+	nsfs := nsStore.Spec.NSFS
+
+	if nsfs == nil {
+		return nil
+	}
+
+	//pvcName validation
+	if nsfs.PvcName == "" {
+		return util.NewPersistentError("ConfigurationError", "PvcName must not be empty")
+	}
+
+	//SubPath validation
+	if nsfs.SubPath != "" {
+		path := nsfs.SubPath
+		if len(path) > 0 && path[0] == '/' {
+			return util.NewPersistentError("ConfigurationError", 
+			  fmt.Sprintf("SubPath %s must be a relative path", path))
+		}
+		parts := strings.Split(path, "/")
+		for _, item := range parts {
+			if item == ".." {
+				return util.NewPersistentError("ConfigurationError", 
+			  	  fmt.Sprintf("SubPath %s must not contain '..'", path))
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateNsStoreS3Compatible validates namespacestore S3Compatible type configuration
+func validateNsStoreS3Compatible(nsStore *nbv1.NamespaceStore) error {
+	s3Compatible := nsStore.Spec.S3Compatible
+
+	if s3Compatible == nil {
+		return nil
+	}
+
+	err := validateSignatureVersion(s3Compatible.SignatureVersion, nsStore.Name)
+	if err != nil {
+		return err
+	}
+
+	err = validateEndPoint(&s3Compatible.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateNsStoreIBMCos validates namespacestore IBMCos type configuration
+func validateNsStoreIBMCos(nsStore *nbv1.NamespaceStore) error {
+	IBMCos := nsStore.Spec.IBMCos
+
+	if IBMCos == nil {
+		return nil
+	}
+
+	err := validateSignatureVersion(IBMCos.SignatureVersion, nsStore.Name)
+	if err != nil {
+		return err
+	}
+
+	err = validateEndPoint(&IBMCos.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//SignatureVersion validation, must be empty or v2 or v4
+func validateSignatureVersion(signature nbv1.S3SignatureVersion, nsStoreName string) error {
+	if signature != "" && 
+		signature != nbv1.S3SignatureVersionV2 && 
+		signature != nbv1.S3SignatureVersionV4 {
+		return util.NewPersistentError("ConfigurationError",
+			fmt.Sprintf("Invalid s3 signature version %q for namespace store %q",
+			signature, nsStoreName))
+	}
+	return nil
+}
+
+//Endpoint validation and sets default
+func validateEndPoint(endPointPointer *string) error {
+	endPoint := *endPointPointer
+
+	if endPoint == "" {
+		endPoint = defaultEndPointURI
+	}
+
+	if endPoint == "" {
+		return util.NewPersistentError("ConfigurationError", "endpoint url must be not empty")
```

Comment:
why do you return an error here?

Reply:
In my opinion, the same reason why you asked me to add the additional validation in the MakeExternalConnectionParams.
if someone changes the default URI to empty, raise the error.

Reply:
I think we didn't understand each other. of course you shouldn't have the same validations in `MakeExternalConnection`. 
I did think we should make sure that functions like `getAuthMethod` will validate that the signature version is valid and not just return `v4` in the else clause.

in this case, there is no code between the 2 checks, so the second one is redundant.